pkg/coordinator/sweep.go

```diff
@@ -127,3 +132,129 @@ func ValidateDepositString(depositString string) error {
 	}
 	return nil
 }
+
+func EstimateDepositsSweepFee(
+	tbtcChain tbtc.Chain,
+	btcChain bitcoin.Chain,
+	depositsCount int,
+) error {
+	_, _, perDepositMaxFee, _, err := tbtcChain.GetDepositParameters()
+	if err != nil {
+		return fmt.Errorf("cannot get deposit tx max fee: [%v]", err)
+	}
+
+	fees := make(map[int]int64)
+	var depositsCountKeys []int
+
+	if depositsCount > 0 {
+		depositsCountKeys = append(depositsCountKeys, depositsCount)
+	} else {
+		sweepMaxSize, err := tbtcChain.GetDepositSweepMaxSize()
+		if err != nil {
+			return fmt.Errorf("cannot get sweep max size: [%v]", sweepMaxSize)
+		}
+
+		for i := 1; i <= int(sweepMaxSize); i++ {
+			depositsCountKeys = append(depositsCountKeys, i)
+		}
+	}
+
+	for _, depositsCountKey := range depositsCountKeys {
+		fee, err := estimateDepositsSweepFee(
+			btcChain,
+			depositsCountKey,
+			perDepositMaxFee,
+		)
+		if err != nil {
+			return fmt.Errorf(
+				"cannot estimate fee for deposits count [%v]: [%v]",
+				depositsCountKey,
+				err,
+			)
+		}
+
+		fees[depositsCountKey] = fee
+	}
+
+	err = printDepositsSweepFeeTable(fees)
+	if err != nil {
+		return fmt.Errorf("cannot print fees table: [%v]", err)
+	}
+
+	return nil
+}
+
+func estimateDepositsSweepFee(
+	btcChain bitcoin.Chain,
+	depositsCount int,
+	perDepositMaxFee uint64,
+) (int64, error) {
+	transactionSize, err := bitcoin.NewTransactionSizeEstimator().
+		// 1 P2WPKH main UTXO input.
+		AddPublicKeyHashInputs(1, true).
+		// depositsCount P2WSH deposit inputs.
+		AddScriptHashInputs(depositsCount, depositScriptByteSize, true).
```

Comment:
We also allow for P2SH input, correct? 

P2SH is longer than P2WSH. What it the % margin we can be off with the estimation? Worth making it very clear in `EstimateDepositsSweepFee` docs.

Reply:
The dapp produces P2WSH deposits addresses and I don't think we have a single P2SH made manually atm. Even if we have one P2SH sometimes, that should not have a big impact if we sweep ~5 deposits at the same time. One P2SH input is `~150 vB` bigger than a P2WSH.

Reply:
The estimated transaction size for 10 inputs is 1022 vBytes. Currently, the estimator suggests `satPerVByte = 1` so the fee for the transaction is 1022 sats. If half of those inputs are P2SH, it means we need to add `2 x 150` sat to the fee. `300` sat is not a lot in USD but it is almost 30% more compared to the original fee.

The estimated transaction size for 5 inputs is 566 vBytes. If 2 inputs are P2SH, it means we need to add `2 x 150` sat to the fee. This is more than 50%  compared to the original fee.

Unless I missed something, the difference is quite significant. 

I don't consider it to be a blocker but I think we should:
- Short term (this PR): Warn about it in `EstimateDepositsSweepFee` docs and in the `estimateDepositsSweepFeeCommandDescription`.
- Mid-term (consider next week): in `getDeposits` function of the coordinator, check the type of a deposit having `event.FundingTxHash`, `event.FundingOutputIndex` to say if it is P2SH or P2WSH.
- Long-term (when automating): estimate the cost based on the actual transaction.

Reply:
I see the short-term proposition was already handled in 7eda5686d1e2b32d4403f628dc240a654be3f7ca when I was writing this comment. Let me know please what you think about mid-term and long-term goals and I can open separate issues for them.

Reply:
What really matters is the actual `sat/vbyte` fee of the transaction. Let's use an example:

At the moment, according to https://mempool.space/pl, we need to:
- Pay `~52 sat/vbyte` to include the transaction in the next block
- Pay `~47 sat/vbyte` to include the transaction in the second block
- Pay `~41 sat/vbyte` to include the transaction in the third block

If we assume we have a 10 P2WSH deposit transaction of size `1022 vBytes`, we will propose `1022*52 = 53144 sat` as a total fee to include it in the next block. If it turns out that half of the actual transaction are P2SH deposits, the transaction size will be around `1770 bytes`. We used `53144 sat` as the total fee which gives `53144/1770 = 30 sat/vbyte` as the actual fee. That means the transaction will be included in 4-5 blocks. This still works and won't cause any serious problems despite the transaction will be included a bit later. 

If we want to be super cautious, the mid-term proposition makes perfect sense here. We can print the type of the deposit transaction while listing deposits and, if there are some P2SH ones and we want to propose them to the sweep, we just manually propose a higher fee than the one printed by the CLI.

Regarding the long-term proposition, I don't think we can build an actual transaction in another place than the wallet's sweep action code. In order to obtain a super-accurate estimation, we just automatically check the deposit types (just like in the mid-term proposition), check whether there is a wallet main UTXO (now we always assume it is present), and then use the current `TransactionSizeEstimator`. This will give the same result (+/- 1 byte for signature padding) as building an actual transaction which is really hard in an environment other than wallet action.

Reply:
> At the moment, according to https://mempool.space/pl, we need to:
> 
> * Pay `~52 sat/vbyte` to include the transaction in the next block
> * Pay `~47 sat/vbyte` to include the transaction in the second block
> * Pay `~41 sat/vbyte` to include the transaction in the third block

This is potentially alarming because when I run `./keep-client coordinator estimate-deposits-sweep-fee` against `electrum.blockstream.info:60001` I am always getting 1 sat/vbyte:
```
 deposits count total fee (satoshis) sat/vbyte
              1                  201         1
              2                  292         1
              3                  384         1
              4                  475         1
              5                  566         1
```

> If we assume we have a 10 P2WSH deposit transaction of size `1022 vBytes`, we will propose `1022*52 = 53144 sat` as a total fee to include it in the next block. If it turns out that half of the actual transaction are P2SH deposits, the transaction size will be around `1770 bytes`. We used `53144 sat` as the total fee which gives `53144/1770 = 30 sat/vbyte` as the actual fee. That means the transaction will be included in 4-5 blocks. This still works and won't cause any serious problems despite the transaction will be included a bit later.

This is ~40% decrease compared to the value we should propose:
```
53144 / 1022 = 52 sat/vbyte
53144 / 1770 = 30 sat/vbyte
```

Agreed, it still works in the current market conditions but the % change is quite significant and we can not guarantee this will work correctly a few months from today. 

I will open an issue to track mid-term and long-term work but in the meantime... what's off with the Blockstream's estimation?

Reply:
I am not running integration tests, I am running `./keep-client coordinator estimate-deposits-sweep-fee` on a binary that I have previously built with `make mainnet` pointing to `electrum.blockstream.info:60001` in the config.

Reply:
`electrum.blockstream.info:60001` is testnet. Use `electrum.blockstream.info:50001` for mainnet. Compare https://1209k.com/bitcoin-eye/ele.php?chain=btc and https://1209k.com/bitcoin-eye/ele.php?chain=tbtc

Reply:
🤦🏻